beacon-chain/sync/backfill/batcher.go

```diff
@@ -0,0 +1,172 @@
+package backfill
+
+import (
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+var errMaxBatches = errors.New("backfill batch requested in excess of max outstanding batches")
+var errEndSequence = errors.New("sequence has terminated, no more backfill batches will be produced")
+var errCannotDecreaseMinimum = errors.New("The minimum backfill slot can only be increased, not decreased")
+
+type batchSequencer struct {
+	batcher batcher
+	seq     []batch
+}
+
+// moveMinimum enables the backfill service to change the slot where the batcher will start replying with
+// batch state batchEndSequence (signaling that no new batches will be produced). This is done in response to
+// epochs advancing, which shrinks the gap between <checkpoint slot> and <current slot>-MIN_EPOCHS_FOR_BLOCK_REQUESTS,
+// allowing the node to download a smaller number of blocks.
+func (c *batchSequencer) moveMinimum(min primitives.Slot) error {
+	if min < c.batcher.min {
+		return errCannotDecreaseMinimum
+	}
+	c.batcher.min = min
+	return nil
+}
+
+func (c *batchSequencer) countWithState(s batchState) int {
+	n := 0
+	for i := 0; i < len(c.seq); i++ {
+		if c.seq[i].state == s {
+			n += 1
+		}
+	}
+	return n
+}
+
+func (c *batchSequencer) update(b batch) {
+	done := 0
+	for i := 0; i < len(c.seq); i++ {
+		if b.replaces(c.seq[i]) {
+			c.seq[i] = b
+		}
+		// Assumes invariant that batches complete and update is called in order.
+		// This should be true because the code using the sequencer doesn't know the expected parent
+		// for a batch until it imports the previous batch.
+		if c.seq[i].state == batchImportComplete {
```

Comment:
What happens if `b.state` already has `batchImportComplete`. Should we even allow that?

Reply:
Did you mean should we allow `c.seq[i] = b` if `c.seq[i].state == batchImportComplete`? In that instance `replaces` will return false, so we won't overwrite it.